core/thread/inc/ThreadPool.h

```diff
@@ -0,0 +1,106 @@
+#ifndef THREADPOOL_H_
+#define THREADPOOL_H_
+
+#include "tbb/tbb.h"
+#include "TPool.h"
+
+class ThreadPool: public TPool<ThreadPool> {
+public:
+   explicit ThreadPool(){
+      this->initTBB->initialize();
+   }
+
+   explicit ThreadPool(size_t nThreads){
+      this->initTBB->initialize(nThreads);
+   }
+
+   ~ThreadPool() {
+      this->initTBB->terminate();
+   }
+
+
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   /// \cond
+   template<class F, class INTEGER> auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector <decltype(func(args.front()))>;
+   template<class F, class T> auto Map(F func, std::vector<T> &args) -> std::vector<decltype(func(args.front()))>;
+   // / \endcond
+   using TPool<ThreadPool>::Map;
+
+   template<class T, class R> auto Reduce(const std::vector<T> &objs, R redfunc) -> decltype(redfunc(objs.front(), objs.front()));
+   using TPool<ThreadPool>::Reduce;
+
+private:
+   tbb::task_scheduler_init *initTBB = new tbb::task_scheduler_init(tbb::task_scheduler_init::deferred);
```

Comment:
I would very, *very* much prefer to have this in a source file, rather than exposing this to all #includers (i.e. same with the tbb include directive). You can do that by inheriting from or aggregating something that's not templated and that does the initialization, such that you can move the implementation to a source file.